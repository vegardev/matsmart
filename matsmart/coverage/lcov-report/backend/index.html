
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for backend</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../prettify.css" />
    <link rel="stylesheet" href="../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../index.html">All files</a> backend</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">79.23% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>1194/1507</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">69.49% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>41/59</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">68.96% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>20/29</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">79.23% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>1194/1507</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line medium'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file medium" data-value="db.tsx"><a href="db.tsx.html">db.tsx</a></td>
	<td data-value="63.49" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 63%"></div><div class="cover-empty" style="width: 37%"></div></div>
	</td>
	<td data-value="63.49" class="pct medium">63.49%</td>
	<td data-value="63" class="abs medium">40/63</td>
	<td data-value="50" class="pct medium">50%</td>
	<td data-value="2" class="abs medium">1/2</td>
	<td data-value="50" class="pct medium">50%</td>
	<td data-value="2" class="abs medium">1/2</td>
	<td data-value="63.49" class="pct medium">63.49%</td>
	<td data-value="63" class="abs medium">40/63</td>
	</tr>

<tr>
	<td class="file high" data-value="definitions.tsx"><a href="definitions.tsx.html">definitions.tsx</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="230" class="abs high">230/230</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="230" class="abs high">230/230</td>
	</tr>

<tr>
	<td class="file high" data-value="dummyData.tsx"><a href="dummyData.tsx.html">dummyData.tsx</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="483" class="abs high">483/483</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="483" class="abs high">483/483</td>
	</tr>

<tr>
	<td class="file medium" data-value="testDatabaseCalls.tsx"><a href="testDatabaseCalls.tsx.html">testDatabaseCalls.tsx</a></td>
	<td data-value="65.15" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 65%"></div><div class="cover-empty" style="width: 35%"></div></div>
	</td>
	<td data-value="65.15" class="pct medium">65.15%</td>
	<td data-value="264" class="abs medium">172/264</td>
	<td data-value="50" class="pct medium">50%</td>
	<td data-value="26" class="abs medium">13/26</td>
	<td data-value="72.22" class="pct medium">72.22%</td>
	<td data-value="18" class="abs medium">13/18</td>
	<td data-value="65.15" class="pct medium">65.15%</td>
	<td data-value="264" class="abs medium">172/264</td>
	</tr>

<tr>
	<td class="file medium" data-value="uploadData.tsx"><a href="uploadData.tsx.html">uploadData.tsx</a></td>
	<td data-value="57.6" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 57%"></div><div class="cover-empty" style="width: 43%"></div></div>
	</td>
	<td data-value="57.6" class="pct medium">57.6%</td>
	<td data-value="467" class="abs medium">269/467</td>
	<td data-value="87.09" class="pct high">87.09%</td>
	<td data-value="31" class="abs high">27/31</td>
	<td data-value="66.66" class="pct medium">66.66%</td>
	<td data-value="9" class="abs medium">6/9</td>
	<td data-value="57.6" class="pct medium">57.6%</td>
	<td data-value="467" class="abs medium">269/467</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at 2024-04-25T18:58:33.555Z
            </div>
        <script src="../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../sorter.js"></script>
        <script src="../block-navigation.js"></script>
    </body>
</html>
    