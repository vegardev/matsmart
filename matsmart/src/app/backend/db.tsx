import mysql from "mysql2/promise";

/**
 * Executes a query on the production database.
 * @param {Object} param0 - An object containing the SQL query string and an optional array of values to be inserted into the query.
 * @param {string} param0.query - The SQL query string.
 * @param {any[]} [param0.values] - An array of values to be inserted into the query.
 * @returns {Promise<any>} A promise that resolves to the results of the query.
 * @throws {Error} When there is an error executing the database query.
 */
export async function query({
  query,
  values = [],
}: {
  query: string;
  values?: any[];
}) {
  const dbconnection = await mysql.createConnection({
    host: "mysql.stud.ntnu.no",
    database: "fs_idatt1005_1_bdigsec4_datab",
    user: "fs_idatt1005_1_group_bdigsec4",
    password: "Password",
  });

  try {
    const [results] = await dbconnection.execute(query, values);
    dbconnection.end();
    return results;
  } catch (error) {
    throw Error((error as Error).message);
  }
}

/**
 * Executes a query on the test database.
 * @param {Object} param0 - An object containing the SQL query string and an optional array of values to be inserted into the query.
 * @param {string} param0.query - The SQL query string.
 * @param {any[]} [param0.values] - An array of values to be inserted into the query.
 * @returns {Promise<any>} A promise that resolves to the results of the query.
 * @throws {Error} When there is an error executing the database query.
 */
export async function testquery({
  query,
  values = [],
}: {
  query: string;
  values?: any[];
}) {
  const dbconnetion = await mysql.createConnection({
    host: "mysql.stud.ntnu.no",
    database: "fs_idatt1005_1_bdigsec4_test",
    user: "fs_idatt1005_1_group_bdigsec4",
    password: "Password",
  });

  try {
    const [results] = await dbconnetion.execute(query, values);
    dbconnetion.end();
    return results;
  } catch (error) {
    throw Error((error as Error).message);
  }
}
