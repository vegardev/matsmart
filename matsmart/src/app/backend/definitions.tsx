/* eslint-disable no-unused-vars */
export type CloseToExpire = {
  item_id: number;
  item_name: string;
  expiration_date: Date;
  quantity: number;
  item_type: string;
};

export type Item_database = {
  item_id: number;
  item_name: string;
  item_quantity_type: string;
};

export type Tags = {
  tag_id: number;
  tag_name: string;
};

export type Recipe_Preview = {
  recipe_id: number;
  recipe_name: string;
  recipe_time: number;
  recipe_image: string;
  recipe_tags: string[];
};

export type Recipe = {
  recipe_id: number;
  title: string;
  recipe_method: string;
  recipe_nutritions: string;
  image: string;
  recipe_time: string;
};

export type Recipes_no_content = {
  recipe_id: number;
  title: string;
  content: string;
  image: string;
};

export type Recipe_items = {
  item_id: number;
  recipe_id: number;
  item_quantity: number;
  item_quantity_type: string;
  item_name: string;
};

//expiration date burde nok endres type til Date
export type Inventory_items = {
  inventory_id?: number;
  item_id: number;
  item_quantity: number;
  item_quantity_type: string;
  item_name: string;
  item_location: string;
  expiration_date?: string;
};

/*
-----------
-----------
Husk Ã¥ fiks opp i type greiene senere
-----------
-----------
*/
export type Inventory_items2 = {
  inventory_id?: number;
  item_id?: number;
  item_quantity: number;
  item_quantity_type: string;
  item_name: string;
  item_location?: string;
  expiration_date: string;
};

export type Recipe_Page = {
  recipe_id: number;
  recipe_name: string;
  recipe_time: number;
  recipe_image: string;
  recipe_tags: string[];
  recipe_method: string;
  recipe_ingredients: string[];
  recipe_nutrition: string;
};

export type QuantityDropdownProps = {
  quantity: number;
  setQuantity: (quantity: number) => void;
  quantityType: string;
  setQuantityType: (quantityType: string) => void;
};

export type Shopping_items = {
  item_id: number;
  item_name: string;
  item_quantity: number;
  item_quantity_type: string;
};
