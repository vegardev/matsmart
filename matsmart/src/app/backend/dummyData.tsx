export const tagsDummyData = [
  { tag_id: 1, tag_name: "Italian" },
  { tag_id: 2, tag_name: "Vegetarian" },
  { tag_id: 3, tag_name: "Vegan" },
  { tag_id: 4, tag_name: "Gluten Free" },
  { tag_id: 5, tag_name: "Dairy Free" },
  { tag_id: 6, tag_name: "Nut Free" },
  { tag_id: 7, tag_name: "Seafood" },
  { tag_id: 8, tag_name: "Poultry" },
  { tag_id: 9, tag_name: "Beef" },
  { tag_id: 10, tag_name: "Pork" },
];

export const recipesDummyData = [
  {
    recipe_id: 1,
    recipe_name: "Pasta",
    recipe_time: 30,
    recipe_image:
      "https://images.matprat.no/p8hgg8jg6v-jumbotron/large/fersk_pasta1.jpg",
    recipe_tags: ["Italian", "Vegetarian", "Dairy Free", "Nut Free"],
  },
  {
    recipe_id: 2,
    recipe_name: "Chicken Curry",
    recipe_time: 45,
    recipe_image: "https://images.matprat.no/xftestavbe-related/tablet",
    recipe_tags: ["Indian", "Poultry"],
  },
  {
    recipe_id: 3,
    recipe_name: "Beef Stir Fry",
    recipe_time: 25,
    recipe_image:
      "https://cdn.momsdish.com/wp-content/uploads/2021/07/Steak-Stir-Fry-Recipe-015.jpg",
    recipe_tags: ["Asian", "Beef"],
  },
  {
    recipe_id: 4,
    recipe_name: "Vegan Salad",
    recipe_time: 15,
    recipe_image:
      "https://cdn77-s3.lazycatkitchen.com/wp-content/uploads/2021/05/green-vegan-salad-platter-800x1200.jpg",
    recipe_tags: ["Vegan", "Gluten Free"],
  },
  {
    recipe_id: 5,
    recipe_name: "Seafood Paella",
    recipe_time: 60,
    recipe_image:
      "https://themediterraneanchick.com/wp-content/uploads/2020/09/IMG_0825-1-scaled.jpg",
    recipe_tags: ["Seafood"],
  },
  {
    recipe_id: 6,
    recipe_name: "Chicken Parmesan",
    recipe_time: 45,
    recipe_image:
      "https://hips.hearstapps.com/hmg-prod/images/delish-202102-airfryerchickenparm-184-ls-1612561654.jpg?crop=1xw:0.84375xh;center,top&resize=1200:*",
    recipe_tags: ["Poultry", "Italian"],
  },
  {
    recipe_id: 7,
    recipe_name: "Beef Tacos",
    recipe_time: 30,
    recipe_image:
      "https://danosseasoning.com/wp-content/uploads/2022/03/Beef-Tacos-768x575.jpg",
    recipe_tags: ["Beef", "Mexican"],
  },
];
