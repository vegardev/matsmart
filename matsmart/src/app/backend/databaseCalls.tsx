import { query } from "@/src/app/backend/db";
import {
  Item_database,
  Recipe,
  Recipes_no_content,
  Recipe_items,
  Inventory_items,
  Shopping_items,
} from "@/src/app/backend/definitions";

//Posts to console if you have connection with the database
console.log("MYSQL_HOST:", process.env.MYSQL_HOST);

//test query to see ifthe datase fetching works
export async function fetchDatabaseTest(): Promise<Item_database[]> {
  try {
    const dbquery = await query({
      query: "SELECT * FROM item_database",
      values: [],
    });
    return dbquery as Item_database[];
  } catch (error) {
    throw Error((error as Error).message);
  }
}
//Fetches all data (excluding content) from all recipes in the database
//  Skal utvidde spørring for å ha tags med
//  Burde legges til en check for om id er større og mindre en noen verdier for å kun vise et bestemt antall per side
export async function fetchRecipes(): Promise<Recipes_no_content[]> {
  try {
    const dbquery = await query({
      query: "SELECT recipe_id, title, image FROM recipes",
      values: [],
    });
    return dbquery as Recipes_no_content[];
  } catch (error) {
    throw Error((error as Error).message);
  }
}

//Fetches a single recipe based on a recipe_id value
export async function fetchSingelRecipe(recipe_id: number): Promise<Recipe[]> {
  try {
    const dbquery = await query({
      query: "SELECT * FROM recipes WHERE recipe_id = ?",
      values: [recipe_id],
    });
    return dbquery as Recipe[];
  } catch (error) {
    throw Error((error as Error).message);
  }
}

//Fetces info about items needed for a recipe based on a recipe_id value
export async function fetchRecipeItems(
  recipe_id: number
): Promise<Recipe_items[]> {
  try {
    const dbquery = await query({
      query:
        "SELECT ri.*, id.item_name FROM recipe_items ri RIGHT JOIN item_database id ON (ri.item_id = id.item_id) WHERE ri.recipe_id = ?",
      values: [recipe_id],
    });
    return dbquery as Recipe_items[];
  } catch (error) {
    throw Error((error as Error).message);
  }
}

//Fetches all items in a inventory based on a location value
// Gjorde endring slik at navn på item hentes fra item_database
export async function fetchInventoryItems(
  location: string
): Promise<Inventory_items[]> {
  try {
    const dbquery = await query({
      query:
        "SELECT inventory.*, item_database.item_name FROM inventory, item_database WHERE inventory.item_id = item_database.item_id AND location = ?",
      values: [location],
    });
    return dbquery as Inventory_items[];
  } catch (error) {
    throw Error((error as Error).message);
  }
}

// Gjør full text search på recipe titles i databasen, og returnerer en liste med oppskrifter som matcher
export async function fetchRecipeSuggestions(
  searchQuery: string
): Promise<Recipes_no_content[]> {
  try {
    const dbquery = await query({
      query:
        "SELECT recipe_id, title, image FROM recipes WHERE title LIKE CONCAT('%', ?, '%')",
      values: [searchQuery],
    });
    return dbquery as Recipes_no_content[];
  } catch (error) {
    throw Error((error as Error).message);
  }
}

// Gjør full text search på item names i databasen, og returnerer en liste med items som matcher
export async function fetchGrocerySuggestions(
  searchQuery: string
): Promise<Item_database[]> {
  try {
    const dbquery = await query({
      query:
        "SELECT * FROM item_database WHERE item_name LIKE CONCAT('%', ?, '%')",
      values: [searchQuery],
    });
    return dbquery as Item_database[];
  } catch (error) {
    throw Error((error as Error).message);
  }
}

// Fetcher alle items i shopping_list
// Gjorde endring slik at navn på shopping list items hentes fra item_database
export async function fetchShoppingList(): Promise<Shopping_items[]> {
  try {
    const dbquery = await query({
      query:
        "SELECT idb.item_name, sl.item_quantity, sl.item_quantity_type FROM item_database idb, shopping_list sl WHERE idb.item_id = sl.item_id",
      values: [],
    });
    return dbquery as Shopping_items[];
  } catch (error) {
    throw Error((error as Error).message);
  }
}

// Submitter ny item til item_database
export async function submitGroceryItem(
  item_name: string,
  item_quantity_type: string
): Promise<void> {
  try {
    await query({
      query:
        "INSERT INTO item_database (item_name, item_quantity_type) VALUES (?, ?)",
      values: [item_name, item_quantity_type],
    });
  } catch (error) {
    throw Error((error as Error).message);
  }
}

// Submitter ny item til shopping_list
export async function submitShoppingListItem(
  item_id: number,
  item_quantity: number,
  item_quantity_type: string
): Promise<void> {
  try {
    await query({
      query:
        "INSERT INTO shopping_list (item_id, item_quantity, item_quantity_type) VALUES (?, ?, ?)",
      values: [item_id, item_quantity, item_quantity_type],
    });
  } catch (error) {
    throw Error((error as Error).message);
  }
}

// Fetcher alle grocery items fra item_database
export async function fetchGroceryItems(): Promise<Item_database[]> {
  try {
    const dbquery = await query({
      query: "SELECT * FROM item_database",
      values: [],
    });
    return dbquery as Item_database[];
  } catch (error) {
    throw Error((error as Error).message);
  }
}
