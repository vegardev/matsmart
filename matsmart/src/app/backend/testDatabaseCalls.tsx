/**
 * This file contains functions for making queries to the test database.
 * The structure and functionality of these functions are identical to those in the `databaseCalls.tsx` file.
 * For detailed documentation on how these functions work, please refer to the corresponding functions in [[databaseCalls]].
 */

import { testquery } from "@/src/app/backend/db";
import {
  Item_database,
  Recipe,
  Recipes_no_content,
  Recipe_items,
  Inventory_items,
  CloseToExpire,
  Shopping_items,
  Tags,
} from "@/src/app/backend/definitions";

console.log("MYSQL_HOST:", process.env.MYSQL_HOST);

export async function fetchDatabaseTest(): Promise<Item_database[]> {
  try {
    const dbquery = await testquery({
      query: "SELECT * FROM item_database",
      values: [],
    });
    return dbquery as Item_database[];
  } catch (error) {
    throw Error((error as Error).message);
  }
}

export async function fetchRecipes(): Promise<Recipes_no_content[]> {
  try {
    const dbquery = await testquery({
      query: "SELECT recipe_id, title, image FROM recipes",
      values: [],
    });
    return dbquery as Recipes_no_content[];
  } catch (error) {
    throw Error((error as Error).message);
  }
}

export async function fetchSingleRecipe(recipe_id: number): Promise<Recipe[]> {
  try {
    const dbquery = await testquery({
      query: "SELECT * FROM recipes WHERE recipe_id = ?",
      values: [recipe_id],
    });
    return dbquery as Recipe[];
  } catch (error) {
    throw Error((error as Error).message);
  }
}

export async function fetchRecipeItems(
  recipe_id: number,
): Promise<Recipe_items[]> {
  try {
    const dbquery = await testquery({
      query:
        "SELECT ri.*, id.item_name FROM recipe_items ri RIGHT JOIN item_database id ON (ri.item_id = id.item_id) WHERE ri.recipe_id = ?",
      values: [recipe_id],
    });
    return dbquery as Recipe_items[];
  } catch (error) {
    throw Error((error as Error).message);
  }
}

export async function fetchInventoryItems(
  location: string,
): Promise<Inventory_items[]> {
  try {
    const dbquery = await testquery({
      query:
        "SELECT inventory.*, item_database.item_name FROM inventory, item_database WHERE inventory.item_id = item_database.item_id AND location = ?",
      values: [location],
    });
    return dbquery as Inventory_items[];
  } catch (error) {
    throw Error((error as Error).message);
  }
}

export async function deleteInventoryItem(item_id: number): Promise<void> {
  try {
    await testquery({
      query: "DELETE FROM inventory WHERE item_id = ?",
      values: [item_id],
    });
  } catch (error) {
    throw Error((error as Error).message);
  }
}
export async function deleteShoppingListItem(item_id: number): Promise<void> {
  try {
    await testquery({
      query: "DELETE FROM shopping_list WHERE item_id = ?",
      values: [item_id],
    });
  } catch (error) {
    throw Error((error as Error).message);
  }
}
export async function fetchCloseToExpireItems(): Promise<CloseToExpire[]> {
  try {
    const dbquery = await testquery({
      query:
        "SELECT i.*, id.item_name FROM inventory i INNER JOIN item_database id ON i.item_id = id.item_id WHERE i.expiration_date IN ( SELECT MIN(expiration_date) FROM inventory GROUP BY item_id ) AND i.expiration_date < DATE_ADD(CURDATE(), INTERVAL 5 DAY) ORDER BY i.expiration_date ASC LIMIT 5",
      values: [],
    });
    return dbquery as CloseToExpire[];
  } catch (error) {
    throw Error((error as Error).message);
  }
}
export async function fetchGroceryItems(): Promise<Item_database[]> {
  try {
    const dbquery = await testquery({
      query: "SELECT * FROM item_database",
      values: [],
    });
    return dbquery as Item_database[];
  } catch (error) {
    throw Error((error as Error).message);
  }
}

export async function fetchGrocerySuggestions(
  searchQuery: string,
): Promise<Item_database[]> {
  try {
    const dbquery = await testquery({
      query:
        "SELECT * FROM item_database WHERE item_name LIKE CONCAT('%', ?, '%')",
      values: [searchQuery],
    });
    return dbquery as Item_database[];
  } catch (error) {
    throw Error((error as Error).message);
  }
}
export async function fetchRecentlyAddedItems(): Promise<Inventory_items[]> {
  try {
    const dbquery = await testquery({
      query:
        "SELECT inventory.*, item_database.item_name FROM inventory, item_database WHERE inventory.item_id = item_database.item_id ORDER BY inventory.inventory_id DESC LIMIT 5",
      values: [],
    });
    return dbquery as Inventory_items[];
  } catch (error) {
    throw Error((error as Error).message);
  }
}
export async function fetchRecipeSuggestions(
  searchQuery: string,
): Promise<Recipes_no_content[]> {
  try {
    const dbquery = await testquery({
      query:
        "SELECT recipe_id, title, image FROM recipes WHERE title LIKE CONCAT('%', ?, '%')",
      values: [searchQuery],
    });
    return dbquery as Recipes_no_content[];
  } catch (error) {
    throw Error((error as Error).message);
  }
}
export async function fetchRecommendedRecipes(): Promise<Recipe[]> {
  try {
    const dbquery = await testquery({
      query: `SELECT *
         FROM recipes 
         WHERE NOT EXISTS (
           SELECT 1 FROM recipe_items 
           WHERE recipe_items.recipe_id = recipes.recipe_id 
           AND recipe_items.item_id NOT IN (
             SELECT inventory.item_id FROM inventory
           )
         )`,
    });
    return dbquery as Recipe[];
  } catch (error) {
    throw Error((error as Error).message);
  }
}
export async function fetchShoppingList(): Promise<Shopping_items[]> {
  try {
    const dbquery = await testquery({
      query:
        "SELECT sl.item_id, idb.item_name, sl.item_quantity, sl.item_quantity_type FROM item_database idb, shopping_list sl WHERE idb.item_id = sl.item_id",
      values: [],
    });
    return dbquery as Shopping_items[];
  } catch (error) {
    throw Error((error as Error).message);
  }
}
export async function sortByTag(searchQuery: string): Promise<Tags[]> {
  try {
    const dbquery = await testquery({
      query: "SELECT * FROM tags WHERE tag_name LIKE CONCAT('%', ?, '%')",
      values: [searchQuery],
    });
    return dbquery as Tags[];
  } catch (error) {
    throw Error((error as Error).message);
  }
}

export async function submitInventoryItem(
  item_id: number,
  item_quantity: number,
  item_quantity_type: string,
  location: string,
  expiration_date: Date,
): Promise<void> {
  try {
    await testquery({
      query:
        "INSERT INTO inventory (item_id, item_quantity, item_quantity_type, location, expiration_date) VALUES (?, ?, ?, ?, ?)",
      values: [
        item_id,
        item_quantity,
        item_quantity_type,
        location,
        expiration_date,
      ],
    });
  } catch (error) {
    throw Error((error as Error).message);
  }
}

export async function submitShoppingListItem(
  item_id: number,
  item_quantity: number,
  item_quantity_type: string,
): Promise<void> {
  try {
    await testquery({
      query:
        "INSERT INTO shopping_list (item_id, item_quantity, item_quantity_type) VALUES (?, ?, ?)",
      values: [item_id, item_quantity, item_quantity_type],
    });
  } catch (error) {
    throw Error((error as Error).message);
  }
}
export async function updateShoppingListItem(
  item_id: number,
  itemQuantity: number,
): Promise<void> {
  try {
    await testquery({
      query: "UPDATE shopping_list SET item_quantity = ? WHERE item_id = ?",
      values: [itemQuantity, item_id],
    });
  } catch (error) {
    throw Error((error as Error).message);
  }
}
