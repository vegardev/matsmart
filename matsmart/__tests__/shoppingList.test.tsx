import React from "react";
import ShoppingList from "@/src/app/shoppinglist/page";
import { shoppingListDummyData } from "@/src/app/backend/dummyData";
import { render, waitFor, fireEvent } from "@testing-library/react";
import fetch from "jest-fetch-mock";

describe("Shopping List page", () => {
  beforeEach(() => {
    fetch.resetMocks();
    fetch.mockResponse(JSON.stringify(shoppingListDummyData));
  });

  it("renders correctly", async () => {
    const { asFragment } = render(<ShoppingList />);
    await waitFor(() => {
      expect(asFragment()).toMatchSnapshot();
    });
  });

  it("fetches data and updates state", async () => {
    const { getByText } = render(<ShoppingList />);
    await waitFor(() => {
      expect(fetch).toHaveBeenCalledTimes(2);
      expect(getByText(shoppingListDummyData[0].item_name)).toBeInTheDocument();
    });
  });

  it("handles checkbox change", async () => {
    const { findAllByTestId } = render(<ShoppingList />);
    const checkboxInputs = await findAllByTestId("checkbox-input");
    fireEvent.click(checkboxInputs[1]);
    await waitFor(() => {
      expect(checkboxInputs[1]).toBeChecked();
    });
  });

  it("handles quantity change", async () => {
    const { findAllByTestId } = render(<ShoppingList />);
    const quantityInputs = await findAllByTestId("quantity-input");
    fireEvent.change(quantityInputs[0], { target: { value: 2 } });
    await waitFor(() => {
      expect(quantityInputs[0]).toHaveValue(2);
    });
  });

  it("handles expiry date change", async () => {
    const { findAllByTestId } = render(<ShoppingList />);
    const checkboxes = await findAllByTestId("checkbox-input");
    fireEvent.click(checkboxes[0]);
    const dateInputs = await findAllByTestId("expiry-date-input");
    fireEvent.change(dateInputs[0], { target: { value: "2025-12-31" } });
    await waitFor(() => {
      expect(dateInputs[0]).toHaveValue("2025-12-31");
    });
  });

  it("handles location change", async () => {
    const { findAllByTestId, findByTestId } = render(<ShoppingList />);
    const checkboxes = await findAllByTestId("checkbox-input");
    fireEvent.click(checkboxes[0]);
    const locationInput = await findByTestId("location-input");
    fireEvent.change(locationInput, { target: { value: "kjøleskap" } });
    await waitFor(() => {
      expect(locationInput).toHaveValue("kjøleskap");
    });
  });

  it("handles delete", async () => {
    const { findAllByTestId } = render(<ShoppingList />);
    const deleteButtons = await findAllByTestId("delete-button");
    fireEvent.click(deleteButtons[0]);
    await waitFor(() => {
      expect(deleteButtons[0]).toBeInTheDocument();
    });
  });
});
