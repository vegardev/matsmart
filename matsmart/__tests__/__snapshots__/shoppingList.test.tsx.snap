// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shopping List page renders correctly 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col"
  >
    <h1
      class="mb-4 text-xl md:text-2xl"
    >
      Shopping list
    </h1>
    <div
      class="mt-4 flex flex-row items-baseline justify-items-center gap-2 md:mt-8 max-w-3xl"
    >
      <div
        class="relative flex flex-col"
      >
        <div
          class="flex flex-col flex-1 flex-shrink-0"
        >
          <label
            class="sr-only"
            for="search"
          >
            Search
          </label>
          <input
            autocomplete="off"
            class="peer block w-full border border-gray-200 py-[9px] pl-3 text-sm outline-2 placeholder:text-gray-500 rounded-md"
            id="search"
            placeholder="Search for grocery items..."
            value=""
          />
        </div>
      </div>
      <div
        class="flex flex-row"
      >
        <input
          class="rounded-l-md border border-gray-200 py-[9px] pl-3 text-sm outline-2"
          min="0.5"
          placeholder="1.5"
          step="0.5"
          type="number"
          value="1"
        />
        <select
          class="rounded-r-md border border-gray-200 bg-gray-50 py-[9px] text-sm outline-2"
          data-testid="measure"
          id="measure"
          name="measure"
        >
          <option
            value="stk"
          >
            stk.
          </option>
          <option
            value="liter"
          >
            L
          </option>
          <option
            value="desiliter"
          >
            dL
          </option>
          <option
            value="milliliter"
          >
            mL
          </option>
          <option
            value="kilogram"
          >
            kg
          </option>
          <option
            value="gram"
          >
            g
          </option>
          <option
            value="teskje"
          >
            ts
          </option>
          <option
            value="spiseskje"
          >
            ss
          </option>
          <option
            value="kopp"
          >
            kopp
          </option>
        </select>
      </div>
      <button
        class="rounded-md border border-gray-200 bg-gray-50 p-2 text-sm outline-2 text-center"
        data-testid="add-to-shopping-list-button"
      >
        Add to shopping list
      </button>
    </div>
    <table
      class="min-w-full divide-y rounded-md divide-gray-200 mt-4"
    >
      <thead
        class="bg-gray-50"
      >
        <tr>
          <th
            class="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider"
          >
            Groceries
          </th>
          <th
            class="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider"
          >
            Amount
          </th>
          <th
            class="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider"
          >
            Bought?
          </th>
          <th
            class="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider"
          >
            Remove?
          </th>
        </tr>
      </thead>
      <tbody
        class="bg-white divide-y divide-gray-200"
      />
    </table>
  </div>
</DocumentFragment>
`;
